File: android/beagle/src/test/java/br/com/zup/beagle/android/imagedownloader/DefaultImageDownloaderTest.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.android.imagedownloader

import android.widget.ImageView
import br.com.zup.beagle.android.extensions.once
import br.com.zup.beagle.android.setup.BeagleEnvironment
import br.com.zup.beagle.android.widget.RootView
import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkObject
import io.mockk.verify
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class DefaultImageDownloaderTest {

    private lateinit var defaultImageDownloader: DefaultImageDownloader
    private val imageView: ImageView = mockk(relaxed = true)
    private val rootView: RootView = mockk(relaxed = true)

    @BeforeEach
    fun setUp() {
        mockkObject(BeagleEnvironment)
        every { BeagleEnvironment.beagleSdk } returns mockk(relaxed = true)
        defaultImageDownloader = DefaultImageDownloader()
    }

    @Test
    fun `GIVEN url WHEN download image THEN call post image`()  {
        // Given
        val url = "https://vitafelice.com.br/wp-content/uploads/2019/01/beagle.jpg"

        // When
        defaultImageDownloader.download(url, imageView, rootView)

        // Then
        verify(exactly = once()) { imageView.post(any()) }
    }
}

File: flutter/beagle/test/default/default_image_downloader_test.dart
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import 'dart:convert';
import 'dart:typed_data';

import 'package:beagle/default/default_image_downloader.dart';
import 'package:beagle/interface/http_client.dart';
import 'package:beagle/model/response.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/mockito.dart';

class MockHttpClient extends Mock implements HttpClient {}

void main() {
  group('Given a DefaultBeagleImageDownloader', () {
    group(
        'When downloadImage is called and http response is different than 200',
        () {
      test('Then it should throw BeagleImageDownloaderException', () {
        final fakeResponse = Response(404, '', {}, Uint8List(0));
        final httpClientMock = MockHttpClient();
        when(httpClientMock.sendRequest(any)).thenAnswer((invocation) {
          return Future<Response>.value(fakeResponse);
        });
        const url = 'an url';
        final imageDownloader =
            DefaultBeagleImageDownloader(httpClient: httpClientMock);

        expect(() => imageDownloader.downloadImage(url),
            throwsA(isInstanceOf<BeagleImageDownloaderException>()));
      });
    });

    group('When downloadImage is called and downloaded image file is empty',
        () {
      test('Then it should throw an Exception', () {
        final fakeResponse = Response(200, '', {}, Uint8List(0));
        final httpClientMock = MockHttpClient();
        when(httpClientMock.sendRequest(any)).thenAnswer((invocation) {
          return Future<Response>.value(fakeResponse);
        });
        const url = 'an url';
        final imageDownloader =
            DefaultBeagleImageDownloader(httpClient: httpClientMock);

        expect(() => imageDownloader.downloadImage(url),
            throwsA(isInstanceOf<Exception>()));
      });
    });

    group('When downloadImage is called', () {
      test('Then it should return an Uint8List', () async {
        final fakeImage = base64Decode(
            'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==');
        final fakeResponse = Response(200, '', {}, fakeImage);
        final httpClientMock = MockHttpClient();
        when(httpClientMock.sendRequest(any)).thenAnswer((invocation) {
          return Future<Response>.value(fakeResponse);
        });
        const url = 'an url';
        final imageDownloader =
            DefaultBeagleImageDownloader(httpClient: httpClientMock);

        final downloadedImage = await imageDownloader.downloadImage(url);

        expect(downloadedImage, fakeImage);
      });
    });
  });
}
